rust/connlib/tunnel/src/dns.rs

```diff
@@ -161,3 +162,66 @@ where
     }
     Some(answer_builder.finish())
 }
+
+fn resource_from_question<N: ToDname>(
+    resources: &ResourceTable<ResourceDescription>,
+    question: &Question<N>,
+) -> Option<ResolveStrategy<ResourceDescription, DnsQueryParams>> {
+    let name = ToDname::to_cow(question.qname()).to_string();
+    let qtype = question.qtype();
+    match qtype {
+        Rtype::A | Rtype::Aaaa => resources
+            .get_by_name(&name)
+            .cloned()
+            .map(ResolveStrategy::LocalResponse)
+            .unwrap_or(ResolveStrategy::new(name, qtype))
+            .into(),
+        Rtype::Ptr => {
+            let mut dns_parts = name.split('.').rev();
+            if !dns_parts
+                .next()
+                .is_some_and(|d| d == REVERSE_DNS_ADDRESS_END)
+            {
+                return None;
+            }
+            let ip: IpAddr = match dns_parts.next() {
+                Some(REVERSE_DNS_ADDRESS_V4) => {
+                    let mut ip = [0u8; 4];
+                    for i in ip.iter_mut() {
+                        *i = dns_parts.next()?.parse().ok()?;
+                    }
+                    ip.into()
+                }
+                Some(REVERSE_DNS_ADDRESS_V6) => {
+                    let mut ip = [0u8; 16];
+                    for i in ip.iter_mut() {
+                        *i = u8::from_str_radix(
+                            &format!("{}{}", dns_parts.next()?, dns_parts.next()?),
+                            16,
+                        )
+                        .ok()?;
+                    }
+                    ip.into()
+                }
+                _ => return None,
+            };
+
+            if dns_parts.next().is_some() {
+                return None;
+            }
+
+            resources
+                .get_by_ip(ip)
+                .cloned()
+                .map(ResolveStrategy::LocalResponse)
+                .unwrap_or(ResolveStrategy::new(name, qtype))
+                .into()
+        }
+        _ => None,
+    }
+}
+
+pub(crate) fn as_dns_message(pkt: &IpPacket) -> Option<TrustDnsMessage> {
+    let datagram = pkt.as_udp()?;
+    TrustDnsMessage::from_vec(datagram.payload()).ok()
+}
```

Comment:
What happens if this a TCP packet? I'm guessing we intercept UDP and then reply with UDP, but how would that work for the larger requests? Can test with this and see under what cases DNS resolution is slow / fails.

Reply:
Right now, the TCP packet will be dropped. To support that use-case we need to have our own TCP state-machine to work, we can't simply intercept TCP request, I'd defer implementing this for 1.0 or even later.

But I think as it's this will most likely work since:
* According to [RFC5966](https://www.rfc-editor.org/rfc/rfc5966#section-4) every request *SHOULD* be initiated with a UDP request. since the `pkt` here is the original packet that trigger the request, this should still be correct.
* We can still query/recieve TCP responses, since the packet we create a separated connection to the DNS server.
* That means that we can still use DNSSEC(as much as it's supported by our resolver lib) for querying the server
* There's the edge-case of making zone-transfers when dns forwarding is active, this could fail but it's also unlikely that it will happen.
* Clients should always disable DNSSEC for the sentry DNS address

So, there could be some unlikely edge case of some records we will send the truncated response but it's not likely under normal operation.